content/post/deduplicating_pga_with_apollo.md

```diff
@@ -0,0 +1,398 @@
+---
+author: romain
+date: 2018-10-04
+title: "Deduplicating files in Public Git Archive"
+image: /post/deduplicating_pga_with_apollo/smile.png
+description: "We describe how we ran apollo on PGA, in order to find communities of duplicate files."
+categories: ["science", "technical"]
+draft: false
+---
+
+
+This summer, we [announced](../announcing-pga) the release of `Public Git Archive`,
+a dataset with 3TB of Git data from the most starred repositories on GitHub.
+Now it's time to tell how we tried to deduplicate files in the latest revision
+of the repositories in PGA using our research project for code deduplication, [src-d/apollo](https://github.com/src-d/apollo). Before diving deep, let's quickly see why we created it.
+To the best of our knowledge, the only efforts to detect code clones at massive scale have been 
+made by Lopes et. al., who  leveraged a huge corpus of over 428 million files in 4 languages to map code clones on GitHub ([DéjàVu project](http://mondego.ics.uci.edu/projects/dejavu/)).
+They relied on syntactic features, i.e. identifiers (`my_list`, `your_list`, ...) and literals (`if`, `for`, ...), 
+to compute the similarity between a pair of files. PGA has fewer files in the latest (HEAD) revision 
+- 54 million, and we did not want to give our readers a *DéjàVu* by repeating the same analysis. 
+So we aimed at something  different: not only copy-paste between files, but also 
+involuntary rewrites of the same abstractions. Thus we extracted and used semantic features 
+from [Unified Abstract Syntax Trees](https://docs.sourced.tech/babelfish/uast/uast-specification).
+
+[Take me directly to the cool pics, screw the math.](#embedded-dependencies)
+
+## Moon shot 
+ 
+apollo's deduplication pipeline is a 3-step process:
+
+1. Extract [bags of features](http://www.cs.unc.edu/~lazebnik/spring09/lec18_bag_of_features.pdf) from each file and apply [TF-IDF](https://en.wikipedia.org/wiki/Tf%E2%80%93idf) to keep only the most relevant items (feature extraction step).
```

Comment:
I would write "term frequency–inverse document frequency ([TF-IDF](link)" this first time the term appears.